go/vt/vtgate/planbuilder/operators/insert.go

```diff
@@ -111,12 +111,221 @@ func createOperatorFromInsert(ctx *plancontext.PlanningContext, ins *sqlparser.I
+	for _, expr := range compExprs {
+		if expr == nil {
+			continue
+		}
+		if outputExpr == nil {
+			outputExpr = expr
+			continue
+		}
+		outputExpr = &sqlparser.OrExpr{
+			Left:  outputExpr,
+			Right: expr,
+		}
+	}
+	return outputExpr
+}
+
+func pkCompExpression(vTbl *vindexes.Table, ins *sqlparser.Insert, rows sqlparser.Values) *sqlparser.ComparisonExpr {
+	if len(vTbl.PrimaryKey) == 0 {
+		return nil
+	}
+	type pComp struct {
+		idx int
+		def sqlparser.Expr
+	}
+	var pIndexes []pComp
+	var pColTuple sqlparser.ValTuple
+	for _, pCol := range vTbl.PrimaryKey {
+		var def sqlparser.Expr
+		idx := ins.Columns.FindColumn(pCol)
+		if idx == -1 {
+			def = findDefault(vTbl, pCol)
+			if def == nil {
+				// If default value is empty, nothing to compare as it will always be false.
+				return nil
+			}
+		}
+		pIndexes = append(pIndexes, pComp{idx, def})
+		pColTuple = append(pColTuple, sqlparser.NewColName(pCol.String()))
+	}
+
+	var pValTuple sqlparser.ValTuple
+	for _, row := range rows {
+		var rowTuple sqlparser.ValTuple
+		for _, pIdx := range pIndexes {
+			if pIdx.idx == -1 {
+				rowTuple = append(rowTuple, pIdx.def)
+			} else {
+				rowTuple = append(rowTuple, row[pIdx.idx])
+			}
+		}
+		pValTuple = append(pValTuple, rowTuple)
+	}
+	return sqlparser.NewComparisonExpr(sqlparser.InOp, pColTuple, pValTuple, nil)
+}
+
+func findDefault(vTbl *vindexes.Table, pCol sqlparser.IdentifierCI) sqlparser.Expr {
+	for _, column := range vTbl.Columns {
+		if column.Name.Equal(pCol) {
+			return column.Default
+		}
+	}
+	panic(vterrors.VT03014(pCol.String(), vTbl.Name.String()))
+}
+
+func uniqKeyCompExpressions(vTbl *vindexes.Table, ins *sqlparser.Insert, rows sqlparser.Values) (comps []*sqlparser.ComparisonExpr, err error) {
+	noOfUniqKeys := len(vTbl.UniqueKeys)
+	if noOfUniqKeys == 0 {
+		return nil, nil
+	}
+
+	type uComp struct {
+		idx int
+		def sqlparser.Expr
+	}
+
+	type uIdx struct {
+		Indexes [][]uComp
+		uniqKey sqlparser.Exprs
+	}
+
+	allIndexes := make([]uIdx, 0, noOfUniqKeys)
+	allColTuples := make([]sqlparser.ValTuple, 0, noOfUniqKeys)
+	for _, uniqKey := range vTbl.UniqueKeys {
+		var uIndexes [][]uComp
+		var uColTuple sqlparser.ValTuple
+		skipKey := false
+		for _, expr := range uniqKey {
+			var offsets []uComp
+			col, isCol := expr.(*sqlparser.ColName)
+			if isCol {
+				var def sqlparser.Expr
+				idx := ins.Columns.FindColumn(col.Name)
+				if idx == -1 {
+					def = findDefault(vTbl, col.Name)
+					if def == nil {
+						// default value is empty, nothing to compare as it will always be false.
+						skipKey = true
+						break
+					}
+				}
+				offsets = append(offsets, uComp{idx, def})
+			} else {
+				err = sqlparser.Walk(func(node sqlparser.SQLNode) (kontinue bool, err error) {
+					col, ok := node.(*sqlparser.ColName)
+					if !ok {
+						return true, nil
+					}
+					var def sqlparser.Expr
+					idx := ins.Columns.FindColumn(col.Name)
+					if idx == -1 {
+						def = findDefault(vTbl, col.Name)
```

Comment:
here we are at 6 levels of indentation. My brain has a hard time following things deeper than 2 or 3 levels